### PR TITLE
Geoip filter

### PR DESCRIPTION
**Pull request by nezdolik**
This change introduces  IP Geolocation filter. Implementation according to [design doc](https://docs.google.com/document/d/1klskyqKxgv5xBq1Xx38x5i5pUx7dkSztaSaoECJovJA/edit?usp=sharing).

Commit Message: IP Geolocation filter
Additional Description:
Risk Level: Low (new extension)
Testing: unit tests
Docs Changes: Done
Release Notes: tbd
Platform Specific Features: NA
Fixes moby/moby#23736

### CODE REVIEW

File: api/envoy/extensions/filters/http/geoip/v3/geoip.proto
Comment (nezdolik) [inline api/envoy/extensions/filters/http/geoip/v3/geoip.proto:62]: todo rename to driver

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [inline source/extensions/filters/http/geoip/geoip_filter.cc:76]: todo extract into method

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [inline source/extensions/filters/http/geoip/geoip_filter.cc:36]: todo function param can be reduced to string

**Comment by nezdolik** [comment]
@mattklein123 @alyssawilk wdyt if we split work on geolocation filter into 2 PRs? First PR contains the filter implementation and second PR will contain implementation of geolocation provider (Maxmind). Or would you prefer to have all changes in one large PR? Am proposing to split, as current PR is already big.

**Comment by alyssawilk** [comment]
throwing over to raven for first pass review.  Not sure who will be assigned to second pass but I think I can speak for all of us in saying splitting up into smaller PRs is generally helpful :-)

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (ravenblackx) [inline source/extensions/filters/http/geoip/geoip_filter.cc:92]: Are we confident that this is going to be fast enough to not leave the worker thread blocked on I/O for any significant amount of time? It seems likely that this would be risky to extend to other geoip sources for that reason. Might it be better to implement with callbacks and handle asynchronously?

File: test/extensions/filters/http/geoip/config_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/config_test.cc:59]: This is *okay*, but in the event of a test failure it would have completely unhelpful output. Consider defining a bunch of matchers (I think they could be one-liners) instead of one globby function, and doing 
```
MATCHER_P(HasUseXff, expected, "") {return GeoipFilterPeer::useXff(*arg);}

EXPECT_CALL(filter_callback, addStreamDecoderFilter(AllOf(
    HasUseXff(false),
    HasXffNumTrustedHops(0),
    [etc.]
)));
```

Makes even more sense when you can also reuse this in the other `has_expected_config` case[s] below.

File: source/extensions/filters/http/geoip/geoip_provider_config.h
Comment (ravenblackx) [inline source/extensions/filters/http/geoip/geoip_provider_config.h:25]: Seems unsafe for all these functions to be returning references - if there's any kind of dynamic database reload possible, or if the results for an implementation aren't all held in memory at once, the reference could become invalid before it's used. (References wouldn't work with an RPC-based provider, for example.)

**Review by ravenblackx — CHANGES_REQUESTED** [review]

File: api/envoy/extensions/filters/http/geoip/v3/maxmind.proto
Comment (ravenblackx) [inline api/envoy/extensions/filters/http/geoip/v3/maxmind.proto:1]: This file should go in a location that matches its type and the path to the implementation. For example, extensions that go in the cache filter live in `envoy/extensions/http/cache` not in `envoy/extensions/filters/http/cache`.

Since this provider is not http specific at all, I would think `envoy/extensions/geoip_providers` would be the place.

**Comment by ravenblackx** [comment]
> throwing over to raven for first pass review. Not sure who will be assigned to second pass but I think I can speak for all of us in saying splitting up into smaller PRs is generally helpful :-)

Agreeing that maxmind in a separate PR makes sense. I get why maxmind.proto is in this one though, because it's doing cross-referencing and you don't want to need API reviewers to act on the same file twice. It sure would be nice if we could land the API as a first review, with no implementation at all, and then do the implemention afterwards, but those integration checks hate that idea.

File: api/envoy/extensions/filters/http/geoip/v3/geoip.proto
Comment (ravenblackx) [inline api/envoy/extensions/filters/http/geoip/v3/geoip.proto:18]: After you fix the current docs CI failure, I think you're going to run into one here as well - you must add that doc file if you're including a reference to it.

File: source/extensions/filters/http/geoip/geoip_provider_config.h
Comment (nezdolik) [reply source/extensions/filters/http/geoip/geoip_provider_config.h]: good point, missed it

**Comment by nezdolik** [comment]
Thanks for the detailed review @ravenblackx,  i have switched filter code to an async version. The doc checks will still fail on missing documentation for maxmind extension. I will fix it shortly.

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [reply source/extensions/filters/http/geoip/geoip_filter.cc]: Agree, async fits much better for this use case

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [inline source/extensions/filters/http/geoip/geoip_filter.cc:151]: For now i think those two flags are enough to track the state of lookups completion per request. But i may have missed some edge cases

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (ravenblackx) [inline source/extensions/filters/http/geoip/geoip_filter.cc:146]: I think this doesn't work great with asynchronous callbacks - num_lookups_to_perform_ being updated by the callback directly means it would need to be mutex-guarded or atomic to prevent overlapping operations occurring on different threads. There's also an issue that the filter 'this' can potentially be destroyed while an operation is in flight.

You could resolve this (and other raciness issues so you'd no longer need `all_lookups_invoked_`) by having the callback go through the dispatcher on decoder_callbacks_, and capture a weak_ptr to something owned by the filter. Using the dispatcher effectively makes the filter single-threaded again (i.e. the filter part of the callback can't be called while decodeHeaders is still processing), and capturing a weak_ptr means you can abandon the rest of the callback if the filter has been destroyed before the callback is called.

I don't think there's a way to do this safely where the driver takes only a reference to the callback object - it must take a shared_ptr or a std::function so that it's safe for the filter to be destroyed without causing the driver's async callback to lose its necessary context.

My suggestion would be to pass the driver a unique_ptr to an additional LookupCallbacks implementation that contains a reference to the dispatcher and a weak_ptr that leads back to the filter. When called, that object could then `dispatcher_->dispatch([the_weak_ptr, callback_params]() {if (auto p = the_weak_ptr.lock()) p->onLookupComplete(callback_params);})`

File: test/extensions/filters/http/geoip/geoip_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/geoip_filter_test.cc:144]: Could you make dummy_driver_ also SaveArg<2> (or 3), the callback, and then instead of directly calling onLookupComplete, call it through the captured callback? That would get you better coverage of how the filter actually works. Current testing doesn't verify that the driver gets a correct callback, nor that calling the driver's callback does anything.

(Might make more sense to not use SaveArg, and instead make a custom function that captures all the args into a specified dummy_driver_ member variable struct, so you can save the different callbacks to different slots without having to copy-paste 8 near-duplicate sets of SaveArg.)

File: test/extensions/filters/http/geoip/geoip_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/geoip_filter_test.cc:273]: Combination of the other comments - please add a test `NoCrashIfFilterDestroyedBeforeCallbackCalled`, in which filter_->decodeHeaders is called and queues up some driver callbacks, then `filter_.clear()`, then call the callbacks.

File: source/extensions/filters/http/geoip/geoip_filter.h
Comment (ravenblackx) [inline source/extensions/filters/http/geoip/geoip_filter.h:119]: I think the Envoy way to do this now is
`OptRef<Http::RequestHeaderMap> request_headers_;`

File: test/extensions/filters/http/geoip/geoip_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/geoip_filter_test.cc:150]: Overriden -> Overridden

File: test/extensions/filters/http/geoip/geoip_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/geoip_filter_test.cc:73]: Could you make this a `MATCHER_P`? (It would make it easier for the next person to find and move into a common place, if you don't want to put it somewhere common already, and you get better line numbers on failure with a proper EXPECT_THAT rather than a helper function.)

Also, right now you'd get an exception thrown on the EXPECT_EQ if the EXPECT_TRUE fails, and not very helpful output. As a MATCHER_P you could easily `*result_listener << "bla bla bla" << request_headers` on each type of failure.

File: test/extensions/filters/network/redis_proxy/command_splitter_impl_test.cc
Comment (ravenblackx) [inline test/extensions/filters/network/redis_proxy/command_splitter_impl_test.cc:76]: This looks like a change that belongs in a different PR.

**Review by ravenblackx — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [reply source/extensions/filters/http/geoip/geoip_filter.cc:146]: I may have misinterpreted the docs, was thinking that http filter instance is created per stream/request and that no guarding is needed.

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [reply source/extensions/filters/http/geoip/geoip_filter.cc:134]: agree

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (ravenblackx) [reply source/extensions/filters/http/geoip/geoip_filter.cc:146]: No, you have that part right. The issue is that the driver isn't part of the filter, and (being asynchronous) may have its own thread. So now there's the filter's thread and the driver's thread (or threads! There's multiple calls to the driver in flight at the same time), both/all touching about the same variables. If the callbacks are applied through the dispatcher then they end up back on the filter's thread which therefore would eliminate any need to guard the variables. (But does still need to be careful that the filter instance hasn't been destroyed by the time the dispatcher gets around to dispatching.)

**Comment by wbpcode** [comment]
/wait 

Mart it as waiting. CI need to be resolved and previous comments need to be addressed. Feel free to ping the maintainer when It's readly for a new review.

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (nezdolik) [reply source/extensions/filters/http/geoip/geoip_filter.cc:146]: Yeap, fair. Thanks @ravenblackx and apologies for delay, I have been OOO entire February. Am now catching up and continuing this work.

File: CODEOWNERS
Comment (nezdolik) [inline CODEOWNERS:295]: @mattklein123 i will update the ownership once it's clear who is sponsoring the change

**Comment by ravenblackx** [comment]
I see a CI test failure looking like this:
```
  Expected arg #0: (has use xff true) and (has xff num trusted hops 1) and (has geo header "x-geo-city") and (has geo header "x-geo-region") and (has geo headers size 2) and (has anon geo headers size 1) and (has geo anon header "x-anon-vpn")
           Actual: 0x17ebbf8a2088
```
We get good human-readable output first, but unhelpful "actual". I don't think we can reasonably fix the "actual" because the parameter type being matched is a shared_ptr to a superclass, but I believe you could make this failure output more human-readable by altering the matchers to have additional output, e.g.
```
MATCHER_P(HasUseXff, expected, "") {
  auto filter = std::static_pointer_cast<GeoipFilter>(arg);
  if (GeoipFilterPeer::useXff(*filter) == expected) {
    return true;
  }
  *result_listener << "expected useXff=" << expected << " but was " << GeoipFilterPeer::useXff(*filter);
}
```
(But I'm not sure this will help - this aspect of matchers never behaves quite how I expect.)

**Comment by nezdolik** [comment]
@ravenblackx after updating matchers according to your recommendation the fail case now looks like this:
```test/extensions/filters/http/geoip/config_test.cc:131: EXPECT_CALL(filter_callback, addStreamDecoderFilter( AllOf(HasUseXff(false), HasXffNumTrustedHops(1), HasGeoHeader("x-geo-country"), HasGeoHeader("x-geo-region"), HasGeoHeadersSize(2), HasAnonGeoHeadersSize(1), HasGeoAnonHeader("x-anon-vpn"))))...
  Expected arg #0: (has use xff false) and (has xff num trusted hops 1) and (has geo header "x-geo-country") and (has geo header "x-geo-region") and (has geo headers size 2) and (has anon geo headers size 1) and (has geo anon header "x-anon-vpn")
           Actual: 16-byte object <90-30 DF-3F FE-06 00-00 80-30 DF-3F FE-06 00-00>, expected useXff=0 but was 1
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

File: api/envoy/extensions/filters/http/geoip/v3/geoip.proto
Comment (ravenblackx) [inline api/envoy/extensions/filters/http/geoip/v3/geoip.proto:72]: Why is this in a oneof? An Any type is kind of an implicit version of the same thing.

File: source/extensions/filters/http/geoip/geoip_filter.cc
Comment (ravenblackx) [inline source/extensions/filters/http/geoip/geoip_filter.cc:95]: I think there's no way state_ can be State::Complete here now that you're using the dispatcher, except in unit tests if the mock dispatcher is allowed to just call the callback immediately *and* you have the driver also call its callback immediately. Having a production code branch that's only relevant for tests is best avoided - I'd recommend capturing at least one of the callbacks in unit tests, and calling it after decodeHeaders. That way the code here can be always StopIteration, and the test better reflects actual behavior.

(Otherwise you also need a bunch of unnecessary complexity around *not* calling continueDecoding from the callback if you're still inside decodeHeaders, etc.)

Also, I think with this change `state_` is no longer necessary at all. This is the only place where it's read from, so all the writes to it become meaningless.

File: source/extensions/filters/http/geoip/geoip_filter.h
Comment (ravenblackx) [inline source/extensions/filters/http/geoip/geoip_filter.h:39]: Looks like you need a #include for flat_hash_set, from clangtidy errors.

File: test/extensions/filters/http/geoip/config_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/config_test.cc:54]: Can just return true here, not need header_found, and add some *result_listener output before return false after the loop.

File: test/extensions/filters/http/geoip/config_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/config_test.cc:53]: `if (testing::Matches(expected_header)(header))` is better - using `==` requires expected_header to be a string, versus using `Matches` allows e.g. `EXPECT_THAT(bla, HasGeoHeader(HasSubstr("x")))`

(Same for most of these matchers - if there's an `==` prefer `Matches`)

For single comparisons, prefer `testing::ExplainMatchResult(expected_header, header, result_listener)` which would also allow for any additional output from the other matcher to come through.

File: test/extensions/filters/http/geoip/geoip_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/http/geoip/geoip_filter_test.cc:30]: In general using EXPECT inside a matcher is likely to result in some confusing behaviors on test failure. Better to do `if (fail condition) {*result_listener<<explanation; return false;}`

In this case because it's using get, the expected_header param can't be a matcher, but still worth using `testing::ExplainMatchResult` for the comparison of expected_value so that sub-matchers are an option.

**Review by ravenblackx — CHANGES_REQUESTED** [review]

**Comment by phlax** [comment]
apologies - ive cancelled CI - please hold off pushing until we have progressed the releases that are currently going through

File: test/extensions/filters/http/geoip/geoip_filter_test.cc
Comment (ravenblackx) [reply test/extensions/filters/http/geoip/geoip_filter_test.cc:32]: You changed it to not need the `EXPECT_TRUE`, but forgot to delete the now unnecessary `EXPECT_TRUE`. :)

**Review by ravenblackx — CHANGES_REQUESTED** [review]
Still got a couple of prior comments to resolve too.

**Comment by nezdolik** [comment]
tsan ci failure seems to be unrelated, will merge master and check if it helps

File: CODEOWNERS
Comment (nezdolik) [reply CODEOWNERS:307]: @ravenblackx should i put you instead of Matt as an owner?

Comment (ravenblackx) [reply CODEOWNERS:307]: I guess. It's not really my field, but it would make sense since having reviewed some of it I now have the best context to review the rest of it.

**Review by ravenblackx — APPROVED** [review]

**Review by ravenblackx — APPROVED** [review]

**Comment by nezdolik** [comment]
cc @sc0ttbeardsley i have started working on plugging the geoip provider for this filter in a separate change. If your company uses some other geolocation provider, you will need to submit a patch for it.

**Comment by nezdolik** [comment]
@mattklein123 could you please take the final look? :)

File: api/envoy/extensions/filters/http/geoip/v3/geoip.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/http/geoip/v3/geoip.proto:91]: This should use `TypedExtensionConfig` I think?

File: api/envoy/extensions/filters/http/geoip/v3/geoip.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/http/geoip/v3/geoip.proto:85]: I would refactor this into an XffConfig message, with the presence of the message indicating that XFF should be consulted, and the num_trusted_hops inside of it. This will be a bit clearer from an API perspective.

File: api/envoy/extensions/filters/http/geoip/v3/geoip.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/http/geoip/v3/geoip.proto:45]: Can you specify the concrete values? Is it true/false? Something else? Same for all of the ones that aren't obviously strings.

File: docs/root/configuration/http/http_filters/geoip_filter.rst
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/geoip_filter.rst:45]: This doesn't actually exist. Is this going to be added in some future PR?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks a few high level API comments. Can you also add a release note? Or do we want to wait on that until the maxmind provider is added per my other comment?

/wait

File: docs/root/configuration/http/http_filters/geoip_filter.rst
Comment (nezdolik) [reply docs/root/configuration/http/http_filters/geoip_filter.rst:45]: That's correct. I have initially suggested to split this work into implementing filter itself and then implementing the provider.

**Comment by phlax** [comment]
not sure what is causing the docker fail - but ticket is here #26634 - ive reported upstream also https://github.com/moby/moby/issues/45344

**Comment by nezdolik** [comment]
@mattklein123 regarding release note, i think is better to wait until provider is implemented.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
